post.md

```diff
@@ -0,0 +1,411 @@
+The reason why I grep out lines starting with a dot is
+additional directives which are information for compiler
+rather than actual instructions. 'l' and 'q' at the end of
+instructions marks operand size. It is required only one
+constant->mem write is performed (because there's no way to
+deduce it).
+
+In 64-bit architecture convention changed a little, because
+first 5 parameters (except structures bigger than 8 bytes)
+are passed through registers: `RDI`, `RSI`, `RDX`, `RCX`,
+`R8` and `R9`. As you can see on above code, the return value is put
+into %rax register (unless it's too big), but as the main return
+32-bit value, `EAX` register is actually used.
+
+RBX, RBP, and R12-R15 are considered callee-save, which
+means, that all functions should provide that their values
+will be the same after returning.
+
+# Interrupts and syscalls
+
+Very similar thing are interrupts — they are also kind of
+functions implemented by OS or boot firmware, but they are
+usually called by hardware. When CPU get such interrupt,
+normal execution is stopped and restored when the interrupt is
+handled. Also, CPU itself can generate the interrupt, that's how
+CPU exceptions work (they are issued when some illegal
+instruction is called). There is also `INT` instruction to
+generate the interrupt.
+
+For a long time, this feature was used to provide runtime
+services in BIOS and DOS. In *nix 32-bit OSs (Linux, *BSD)
+it's still used: `int 0x80`. EAX (or it's part) was
+typically specifying system function and other registers
+contained parameters. In 64-bit architecture, there is
+`syscall` instruction instead, which works very similar.
+For example for `read()` function:
+
+```assembly
+	xor %rax,  %rax #sys_read
+	xor %rdi,  %rdi #fd = stdin
+	lea .buff, %rsi #buffer
+	mov $0xff, %rdi #bytes to read
+	syscall
+```
+
+However, those calls are usually called using C wrapper calls
+so the most likely place to find them are shared libraries.
+
+Conclusion
+----------
+
+This brief explanation is probably not enough to code in
+assembly language but will let you understand most of 
+the disassembly of userspace programs. As modern programs make
+much use of shared libraries, those calls used most of the
+time. The good thing is that unless you deal with
+OS/firmware calls you don't need to care about multitasking,
+caching etc. You will probably face strange constructs
+like `call (%rip)`, which doesn't make functional sense, but
+turns out to help CPU execute code faster. Another good news
+is that userspace program is written as though it was only 
+processed running on the machine which simplifies it a lot.
+
+Anyway this should give you good start to understand most of
+assembly code (assuming that you use instruction reference).
+If you deal with obfuscated code, you will proabably need some
+help from dedicated software like IDA.
+
+We are open to help you if some of presented information is
+unclear or you are interested in more detail. Please let us know.
```

Comment:
Anyway, this should give you a good start to understand most of the assembly code (assuming that you use instruction reference). If you deal with obfuscated code, you will probably need some help from dedicated software like IDA.

We are open to help you if some of the presented information is unclear or you are interested in more detail. Please let us know.